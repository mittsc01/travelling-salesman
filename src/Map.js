import React, {useState} from 'react'
import StaticDrawing from './StaticDrawing'
import Helmet from 'react-helmet'
import {MapContainer,TileLayer, useMap} from 'react-leaflet'

export default function Map(props){
  //const [map,setMap] = useState(null)
  const [unit,setUnit] = useState('M')
  //GET API_ENDPOINT/routes/${props.match.params.id}
  const [points, setPoints] = useState([{lat:43.3017, lng:-91.7903},{lat:43.5017, lng:-91.8903}])
  
  


const toggleUnit = () => {
  if (unit==='M'){
    setUnit('K')
  }
  else {
    setUnit('M')
  }
}

function distance(lat1, lon1, lat2, lon2, unit) {
  if ((lat1 == lat2) && (lon1 == lon2)) {
      return 0;
  }
  else {
      var radlat1 = Math.PI * lat1/180;
      var radlat2 = Math.PI * lat2/180;
      var theta = lon1-lon2;
      var radtheta = Math.PI * theta/180;
      var dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
      if (dist > 1) {
          dist = 1;
      }
      dist = Math.acos(dist);
      dist = dist * 180/Math.PI;
      dist = dist * 60 * 1.1515;
      if (unit=="K") { dist = dist * 1.609344 }
      if (unit=="N") { dist = dist * 0.8684 }
      return dist;
  }
}

function computePathLength(points,unit){
  if (points.length === 0){return 0}
  let d = 0 
  for (let i=0;i<points.length-1;i++){
      d+=distance(points[i].lat,points[i].lng,points[i+1].lat,points[i+1].lng,unit)
  }
  return d
}

  return (
    <div className="edit">
      <Helmet>
        <link rel="stylesheet" href="https://unpkg.com/leaflet@1.7.1/dist/leaflet.css"
          integrity="sha512-xodZBNTC5n17Xt2atTPuE1HxjVMSvLVW9ocqUKLsCC5CXdbqCmblAshOMAS6/keqq/sMZMZ19scR4PsZChSR7A=="
          crossorigin="" />
      </Helmet>
      <MapContainer id="map-container" center={{lat: 43.30, lng: -91.79}} zoom={15} scrollWheelZoom={true}>
        <TileLayer
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
        />
        <StaticDrawing points={[...points]}/>
        
      </MapContainer>
      
      <button onClick={toggleUnit}>{unit==='M'?'Metric':'Imperial'}</button>
      <span>{`Route distance: ${computePathLength(points,unit).toFixed(1)} ${unit==='M'?'mi':'km'}`} </span>
    </div>
  );
}
