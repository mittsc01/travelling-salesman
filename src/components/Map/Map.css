
.errand-info {
    text-align: center;
}