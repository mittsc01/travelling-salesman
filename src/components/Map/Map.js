import React, {useState,useEffect} from 'react'
import StaticDrawing from '../StaticDrawing/StaticDrawing'
import Helmet from 'react-helmet'
import {MapContainer,TileLayer, useMap,LayersControl,} from 'react-leaflet'
import RoutesService from '../../services/routes-service'

export default function Map(props){
  //const [map,setMap] = useState(null)
  const [unit,setUnit] = useState('M')
  //GET API_ENDPOINT/routes/${props.match.params.id}
  const [points, setPoints] = useState([])
  const [run,setRun] = useState({})
  
  useEffect( () => {
        
    (async () => {
        const runs =  await RoutesService.getSchedule()
        const theOne = runs.find(run => {
            //console.log(route.id,props.match.params.id)
            return run.id === parseInt(props.match.params.id)
          })
        setRun(theOne)
        //console.log(routes)
        if (runs.length !==0){
            console.log(theOne.route_id)
          const markers = await RoutesService.getPoints(theOne.route_id)
          if (markers.length !==0){
            markers.sort((a,b) => a.index-b.index)
            
          }
          //console.log(markers)
          setPoints(markers)
          
          
        }
         

        
    })()
}, [props])


const toggleUnit = () => {
  if (unit==='M'){
    setUnit('K')
  }
  else {
    setUnit('M')
  }
}

function distance(lat1, lon1, lat2, lon2, unit) {
  if ((lat1 == lat2) && (lon1 == lon2)) {
      return 0;
  }
  else {
      var radlat1 = Math.PI * lat1/180;
      var radlat2 = Math.PI * lat2/180;
      var theta = lon1-lon2;
      var radtheta = Math.PI * theta/180;
      var dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
      if (dist > 1) {
          dist = 1;
      }
      dist = Math.acos(dist);
      dist = dist * 180/Math.PI;
      dist = dist * 60 * 1.1515;
      if (unit=="K") { dist = dist * 1.609344 }
      if (unit=="N") { dist = dist * 0.8684 }
      return dist;
  }
}

function computePathLength(points,unit){
  if (points.length === 0){return 0}
  let d = 0 
  for (let i=0;i<points.length-1;i++){
      d+=distance(points[i].lat,points[i].lng,points[i+1].lat,points[i+1].lng,unit)
  }
  return d
}

  return (
    <div className="edit">
      <Helmet>
        <link rel="stylesheet" href="https://unpkg.com/leaflet@1.7.1/dist/leaflet.css"
          integrity="sha512-xodZBNTC5n17Xt2atTPuE1HxjVMSvLVW9ocqUKLsCC5CXdbqCmblAshOMAS6/keqq/sMZMZ19scR4PsZChSR7A=="
          crossorigin="" />
      </Helmet>
        <h2>{run.title}</h2>
      <MapContainer id="map-container" center={{lat: 43.30, lng: -91.79}} zoom={15} scrollWheelZoom={true}>
       
        <LayersControl position="topright">
        <LayersControl.BaseLayer name="Satellite">
        <TileLayer
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          url='http://mt0.google.com/vt/lyrs=s&x={x}&y={y}&z={z}'
        />
        </LayersControl.BaseLayer>
        <LayersControl.BaseLayer name="Streetview" checked>
        <TileLayer
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" 
        /> 
        </LayersControl.BaseLayer>
        </LayersControl>
       {/* <TileLayer
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          url="https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png" opacity={0.5}
        /> */}
        <StaticDrawing points={[...points]}/>
        
      </MapContainer>
      
      <button onClick={toggleUnit}>{unit==='M'?'Metric':'Imperial'}</button>
      <span>{`Route distance: ${computePathLength(points,unit).toFixed(1)} ${unit==='M'?'mi':'km'}`} </span>
    </div>
  );
}
