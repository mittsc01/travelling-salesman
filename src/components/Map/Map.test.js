import Map from './Map';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom'
it('renders without crashing', () => {
  // first create a DOM element to render the component into
  const div = document.createElement('div');

  // render the component, this is the actual test, if something is wrong it will fail here
  ReactDOM.render(<MemoryRouter><Map /></MemoryRouter>, div);

  // clean up code
  ReactDOM.unmountComponentAtNode(div);
});

it('renders the UI as expected', () => {
  const tree = renderer
    .create(<MemoryRouter><Map /></MemoryRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();  
  });