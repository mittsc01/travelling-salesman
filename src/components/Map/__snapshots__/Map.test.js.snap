// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<div
  className="edit"
>
  <div
    className="errand-info"
  >
    <h2 />
    <p />
  </div>
  <span>
    Route distance: 0.0 mi
     
  </span>
  <div
    id="map-container"
  />
  <button
    onClick={[Function]}
  >
    Metric
  </button>
</div>
`;
